### ISSUE TITLE
Feature request: allow _=_ in user space

### ISSUE THREAD

**Issue by mikeshulman** [agda/agda#5979]
It's a constant source of low-level annoyance to me that Agda doesn't permit using the standard mathematical symbol `=` for propositional equality, and I don't think I'm the only one.  Coq and Lean allow the user to define relations called `=`, using `:=` instead for definitions.  I wouldn't have thought to suggest this could be altered, but some people on the HoTT Zulip thought it might not be impossible for someone to implement an Agda option switching to this notation, so that you could write for instance
```agda
{-# OPTIONS --colon-equals #-}

data _=_ {A : Set} (a : A) : A → Set where
  refl : a = a

rev : {A : Set} {a b : A} → (a = b) → (b = a)
rev refl := refl
```
One would of course have to think about how code using this option would interact with code not using it.  My first guess would be that standard code importing colon-equals code would just have to write `renaming (_=_ to _≡_)` or whatever in order to use the relation `_=_`, and in the other direction colon-equals code importing standard code could choose to write `renaming (_≡_ to _=_)` to be able to use the notation `=` for a previously defined equality.  Of course this means that `_=_` would have to be accepted in a `renaming` declaration whether or not `--colon-equals` is currently in effect.  (And oppositely, if someone somewhere has defined `_:=_` to mean something, then colon-equals code importing such code would have to rename it.)  I'm sure there would be other ways to deal with interoperability, this is just one that occurred to me.

I don't necessarily expect any regular Agda developers to be interested in implementing such an option, but would you consider accepting a patch if someone else implemented it?  Do you see any obvious reasons it would be impossible, difficult, or a bad idea?

**Comment by nad** [comment]
Some thoughts:

* A disadvantage of this kind of feature is that it might make it harder to read Agda code. However, the idea seems to be to make it possible to use a long-standing convention from mathematics, and thus perhaps make the code easier to read for some people.

* If we allow `:=`, then I would also like to allow `≔`. I suggest that we allow arbitrary strings, with some exceptions (whitespace, plus parentheses and other reserved characters). One alternative is to support something like `--definition-string=:=` (but see below).

* Currently the lexer handles `=`, the parser does not care what string is used for this token. I guess one way to implement this feature would be to remove the special treatment of `=`, and to add a special case to the following action:
  https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Parser/LexActions.hs#L251-L253

* If we go with `--definition-string`, then I think this option should at least be supported on the command-line and in `.agda-lib` files. Supporting it in `OPTIONS` pragmas would require changes to how `OPTIONS` pragmas are parsed. One alternative is to ignore the option (but warn about it) if it is given in an `OPTIONS` pragma.

* Agda currently fails to parse the following pieces of code:
  ```agda
  open import M renaming (_=_ to _≡_)
  ```
  ```agda
  open import M renaming (module = to Something)
  ```
  I wonder if it suffices to make the definition of `ImportName_` a little more liberal:
  https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/full/Agda/Syntax/Parser/Parser.y#L1044-L1046

We have previously changed Agda so that one can use `Type` (for instance) instead of `Set`. Are there any other parts of the syntax that should be more customisable? Currently Agda's keywords are all in English, but I can see an argument for allowing keywords in different languages. Instead of having separate command-line options for each piece of customisable syntax one could put everything in a file called (say) `agda.syntax`, placed in the same directory as a library's `.agda-lib` file:
```
=       :=
Set     Typ
module  modul
private privat
where   där
erased  raderad
```
Both this approach and the idea to use `--definition-string` have a disadvantage: it is non-trivial to share definitions between different libraries. Could we instead reuse Agda's library facilities? We could have a builtin type with strings for overridable pieces of syntax, and have an option for telling Agda to use a specific definition:
```
name:   library
depend: agda-syntax
flags:  --syntax=Agda-syntax.HoTT
```
Here `Agda-syntax` should be a module from one of the library's dependencies, and `HoTT` should be a top-level definition of the correct type.

**Comment by mikeshulman** [comment]
I agree that this could make it harder to read code.  There's definitely a tradeoff in customizability versus uniformity.  Allowing renaming of `Set` can also make it harder to read code (I speak from experience!).  I think personally that for this change, like that one, the benefits outweigh the disadvantage, but the disadvantage is definitely real.

I observe that when `Set` is renamed to `Type`, the old meaning of `Set` doesn't go away, and in particular it is not available to be given other meanings.  So there's at least this difference between the two changes.  I suppose one could consider changing the behavior of `Set` to match the proposed behavior of `=`; then for instance one could use `Set` to denote the universe of h-sets in HoTT.

**Comment by nad** [comment]
> I observe that when `Set` is renamed to `Type`, the old meaning of `Set` doesn't go away, and in particular it is not available to be given other meanings.

That is not true:
```agda
{-# OPTIONS --no-import-sorts #-}

open import Agda.Primitive renaming (Set to Type)

data Set : Type where
```

**Comment by mikeshulman** [comment]
Ah, I wasn't aware of the option `--no-import-sorts`.  Sorry for the noise.

**Comment by nicolaikraus** [comment]
It would be great to have such an option in Agda. This would allow us to formalise papers and use the same syntax that the paper uses (which often is `_=_`).

But I worry whether this might lead to confusing error messages? Would it be hard to detect that something doesn't parse simply because the user did/didn't use `--definition-string=:=`?

**Comment by andreasabel** [comment]
In analogy to pattern lambda, a natural symbol to replace `=` would be the right arrow `->`.
In general, making keywords customizable is a bad idea, only escalating the Babylonian language confusion.  Less freedom of expression is good if you want people to understand each other.
Of course, in the case of `=` the idea here seems to move closer to a standard, which is in this context presumed to be Coq rather than Haskell (the departing point for the Agda syntax).

**Comment by nad** [comment]
> In analogy to pattern lambda, a natural symbol to replace `=` would be the right arrow `->`.

I don't want to read or write `Relation A → A → A → Type`.

> Of course, in the case of `=` the idea here seems to move closer to a standard, which is in this context presumed to be Coq rather than Haskell (the departing point for the Agda syntax).

The "standard" is not Coq, but rather mathematics.

**Comment by andreasabel** [comment]
`Relation = \ A → A → A → Type` does not look much better, haha!

**Comment by mikeshulman** [comment]
> In general, making keywords customizable is a bad idea, only escalating the Babylonian language confusion. Less freedom of expression is good if you want people to understand each other.

In general, I agree.  For this reason, I'm wary of the idea to allow the user to specify an arbitrary string for definitions.  I think there would be less potential confusion if there were only two options, `=` and `:=`.

> The "standard" is not Coq, but rather mathematics.

Exactly.  Coq is just following that standard.

>> I don't want to read or write `Relation A → A → A → Type`.
>
> `Relation = \ A → A → A → Type` does not look much better, haha!

Indeed not!  But if it was a bad idea to use `→` for lambdas, it doesn't follow that we should propagate that bad idea to definitions!  (-:O

**Comment by mikeshulman** [comment]
In addition to `≔` (0x2254 COLON EQUALS), other symbols someone might want to use for definition clauses include `≜` (0x225C DELTA EQUAL TO) and `≝` (0x225D EQUAL TO BY DEFINITION).
